File: Day177_2025/Q177_GFG.java
import java.util.*;

class Q177_GFG {
    public int minValue(String s, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        int[] freq = new int[26];

        for (char ch : s.toCharArray()) {
            freq[ch - 'a']++;
        }

        for (int f : freq) {
            if (f > 0) pq.offer(f);
        }

        while (k-- > 0 && !pq.isEmpty()) {
            int top = pq.poll();
            top--;
            if (top > 0) pq.offer(top);
        }

        int ans = 0;
        while (!pq.isEmpty()) {
            int x = pq.poll();
            ans += x * x;
        }

        return ans;
    }
}

File: Day177_2025/Q177_GFG.py
import heapq
from collections import Counter

class Solution:
    def minValue(self, s, k):
        freq = Counter(s)
        max_heap = [-count for count in freq.values()]
        heapq.heapify(max_heap)

        while k > 0 and max_heap:
            top = -heapq.heappop(max_heap)
            top -= 1
            if top > 0:
                heapq.heappush(max_heap, -top)
            k -= 1

        return sum(x * x for x in map(lambda x: -x, max_heap))
